z3tracer/src/parser.rs

```diff
@@ -153,124 +153,133 @@ where
                 let sexp = lexer.read_line()?;
                 let meaning = Meaning { theory, sexp };
                 state.attach_meaning(id, meaning)?;
+                lexer.read_end_of_line()?;
                 Ok(true)
             }
             "[attach-var-names]" => {
                 let id = lexer.read_ident()?;
                 let names = lexer.read_var_names()?;
-                lexer.read_end_of_line()?;
                 state.attach_var_names(id, names)?;
+                lexer.read_end_of_line()?;
                 Ok(true)
             }
             "[inst-discovered]" => {
                 let method = lexer.read_string()?;
-                let key = lexer.read_key()?;
+                let key = lexer.read_fresh_key()?;
                 let quantifier = lexer.read_ident()?;
                 let terms = lexer.read_idents()?;
                 let blame = lexer.read_idents()?;
-                lexer.read_end_of_line()?;
                 let kind = QuantInstantiationKind::Discovered {
                     method,
                     quantifier,
                     terms,
                     blame,
                 };
-                let inst = QuantInstantiation { kind, data: None };
+                let inst = QuantInstantiation {
+                    kind,
+                    data: Vec::new(),
+                };
                 state.add_instantiation(key, inst)?;
+                lexer.read_end_of_line()?;
                 Ok(true)
             }
             "[new-match]" => {
-                let key = lexer.read_key()?;
+                let key = lexer.read_fresh_key()?;
                 let quantifier = lexer.read_ident()?;
                 let trigger = lexer.read_ident()?;
                 let terms = lexer.read_idents()?;
                 let used = lexer.read_matched_terms()?;
-                lexer.read_end_of_line()?;
                 let kind = QuantInstantiationKind::NewMatch {
                     quantifier,
                     trigger,
                     terms,
                     used,
                 };
-                let inst = QuantInstantiation { kind, data: None };
+                let inst = QuantInstantiation {
+                    kind,
+                    data: Vec::new(),
+                };
                 state.add_instantiation(key, inst)?;
+                lexer.read_end_of_line()?;
                 Ok(true)
             }
             "[eq-expl]" => {
                 let id = lexer.read_ident()?;
                 let eq = lexer.read_equality()?;
-                lexer.read_end_of_line()?;
                 state.add_equality(id, eq)?;
+                lexer.read_end_of_line()?;
                 Ok(true)
             }
             "[instance]" => {
                 let key = lexer.read_key()?;
```

Comment:
How do you decide when to use read_key and when to use read_fresh_key? I suppose `[inst-discovered]` and `[new-match]` are for newly found instantiation but `[instance]` is for something already found?

Reply:
That's right: `[instance] xyz` points to the last `xyz` "declared" in a statement `[inst-discovered] xyz` or `[new-match] xyz`.